drgn/helpers/linux/list.py

```diff
@@ -277,3 +284,97 @@ def hlist_for_each_entry(
     type = head.prog_.type(type)
     for pos in hlist_for_each(head):
         yield container_of(pos, type, member)
+
+
+def llist_empty(head: Object) -> bool:
+    """
+    Return whether a llist is empty.
+
+    :param head: ``struct llist_head *``
+    """
+    return head.first.value_() == 0x0
+
+
+def llist_is_singular(head: Object) -> bool:
+    """
+    Return whether an llist has only one element.
+
+    :param head: ``struct llist_head *``
+    """
+    return head.first.value_() != 0x0 and head.first.next.value_() == 0x0
+
+
+def llist_first_entry(head: Object, type: Union[str, Type], member: str) -> Object:
+    """
+    Return the first entry in an llist.
+
+    The list is assumed to be non-empty.
+
+    See also :func:`llist_first_entry_or_null()`.
+
+    :param head: ``struct llist_head *``
+    :param type: Entry type.
+    :param member: Name of llist_node member in entry type.
+    :return: ``type *``
+    """
+    return container_of(head.first, type, member)
+
+
+def llist_first_entry_or_null(
+    head: Object, type: Union[str, Type], member: str
+) -> Object:
+    """
+    Return the first entry in an llist or ``NULL`` if the llist is empty.
+
+    See also :func:`llist_first_entry()`.
+
+    :param head: ``struct llist_head *``
+    :param type: Entry type.
+    :param member: Name of list_node member in entry type.
+    :return: ``type *``
+    """
+    if head.first.value_() == 0x0:
+        return NULL(head.prog_, head.prog_.pointer_type(head.prog_.type(type)))
+    else:
+        return container_of(head.first, type, member)
+
+
+def llist_next_entry(pos: Object, member: str) -> Object:
+    """
+    Return the next entry in an llist.
+
+    :param pos: ``type*``
+    :param member: Name of list_node member in entry type.
+    :return: ``type *``
+    """
+    return container_of(getattr(pos, member).next, pos.type_.type, member)
```

Comment:
None of these functions exist in the kernel. That doesn't necessarily mean that we shouldn't add them as helpers if they're useful, but I'm just curious whether you specifically wanted these or you just added them to duplicate the `list_*` helpers.

Reply:
I am writing some scripts to dump per-CPU call_single_queue ( an llist) and per-CPU call function data. While writing that script the need for these helpers came up. From the above list only llist_next_entry is one which I have not used so far and so far it exists only as a result of duplication from list_* helpers. If needed I can remove this. Other helpers in this list seem helpful to me but I will await your feedback in this regard.